### ISSUE TITLE
send_MP: numbers should not be strings

### ISSUE THREAD

**Issue by dexX7** [mastercoin-MSC/mastercore#103]
Looks like 9c77c604d1730dcf432ae8ba0f1987ff3869e13a came with a side-effect and made it a requirement that the amount is explicitly passed as string. Otherwise a `value is type int, expected str` (or `value is type real, expected str`) is thrown.

This seems to affect RPC which needs a cast to string now, CLI works fine.

Edit: I think this is an issue due to the inconsistency. getbalance_MP for example requires integers and doesn't work with strings.

**Comment by zathras-crypto** [comment]
Hey bud,

We actually require strings now for amounts in RPC calls.  It's less than ideal for sure.

Basically the issue is that json:spirit loses precision at high token counts leading to the wrong amount potentially being encoded (by a few satoshis) - by reading the number in as a string and then doing the conversion internally we work around this issue.

The same is present for amounts in output, they're provided as strings.

This is not a solution, just a work around.  We toyed with modifying json:spirit to use long doubles (so we had more precision) but the general consensus from the team was that floating point should not be used at all, so the string workaround was devised short term.

Thanks
Z

**Comment by m21** [comment]
Has this been resolved guys?

**Comment by dexX7** [comment]
Well yes and no. I still think numbers should be numbers and not strings and it seems dangerous, if you start to parse a string to a number and mix that with localization assumptions ("." vs. "," = decimal point etc.), but that aside:

Big numbers are not only a problem of MC, but I run into such problems as well with (native, no lib) JS, Python or PHP in general.

If this would be possible: be more gentle and error tollerant. If the input is a number, don't throw an exception, if it's within the "accurate range", accept strings nevertheless.. etc. Well, I'm really not sure.. at the same time, I don't like inconsistent behavior..

The later is key. Stick to whatever works, but be consistent. If all "amounts" are strings, then that's it and that's fine. And more telling error messages would be great. I quickly took a look at the JSON specification, but there are no guidelines in this context and written that it's up to an implementation, which range a JSON "number" might be.

Long story short: hashes are actually 256 bit numbers, submitted as hex-encoded string -- this is the closest thing I found. So I think it's fine, if strings are used.
